server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java

```diff
@@ -130,11 +152,27 @@ public synchronized int getNumberOfInFlightFetches() {
      * The ignoreNodes are nodes that are supposed to be ignored for this round, since fetching is async, we need
      * to keep them around and make sure we add them back when all the responses are fetched and returned.
      */
-    public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Set<String> ignoreNodes) {
+    public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Map<ShardId, Set<String>> ignoreNodes) {
         if (closed) {
-            throw new IllegalStateException(shardId + ": can't fetch data on closed async fetch");
+            throw new IllegalStateException(logKey + ": can't fetch data on closed async fetch");
+        }
+
+        if (enableBatchMode == false) {
+            // we will do assertions here on ignoreNodes
+            assert ignoreNodes.size() <= 1 : "Can only have at-most one shard";
```

Comment:
Can we make the assert message more precise, say `Fetching shard data, Can only have at-most one shard`

Reply:
The class is having the pattern to throw Validation failures via assertions. Shouldnt  we should keep the pattern intact for readability and consistency?

Reply:
Make sense. This is the public method, caller can call with any illegal values. Will add illegalStateException